src/sentry_core.c

```diff
@@ -689,3 +710,79 @@ sentry_set_level(sentry_level_t level)
+
+    // TODO(tracing): stuff transaction into the scope
+    bool should_sample = sentry__should_send_transaction(tx_cxt);
+    sentry_value_set_by_key(
+        tx, "sampled", sentry_value_new_bool(should_sample));
+
+    // Avoid having this show up in the payload at all if it doesn't have a
+    // valid value
+    sentry_value_t parent_span
+        = sentry_value_get_by_key_owned(tx_cxt, "parent_span_id");
+    if (sentry_value_get_length(parent_span) > 0) {
+        sentry_value_set_by_key(tx, "parent_span_id", parent_span);
+    }
+    sentry_value_set_by_key(
+        tx, "trace_id", sentry_value_get_by_key_owned(tx_cxt, "trace_id"));
+    sentry_value_set_by_key(
+        tx, "span_id", sentry_value_get_by_key_owned(tx_cxt, "trace_id"));
+    sentry_value_set_by_key(
+        tx, "transaction", sentry_value_get_by_key_owned(tx_cxt, "name"));
+    sentry_value_set_by_key(tx, "start_timestamp",
+        sentry__value_new_string_owned(
+            sentry__msec_time_to_iso8601(sentry__msec_time())));
+
+    sentry_value_decref(tx_cxt);
+
+    return tx;
+}
+
+sentry_uuid_t
+sentry_transaction_finish(sentry_value_t tx)
+{
+    // The sampling decision should already be made for transactions during
+    // their construction. No need to recalculate here. See
+    // `sentry__should_skip_transaction`.
+    sentry_value_t sampled = sentry_value_get_by_key(tx, "sampled");
+    if (!sentry_value_is_null(sampled) && !sentry_value_is_true(sampled)) {
+        SENTRY_DEBUG("throwing away transaction due to sample rate or "
+                     "user-provided sampling value in transaction context");
+        sentry_value_decref(sampled);
+        sentry_value_decref(tx);
+        // TODO(tracing): remove from scope
+        return sentry_uuid_nil();
+    }
+    sentry_value_decref(sampled);
+
+    sentry_value_set_by_key(tx, "type", sentry_value_new_string("transaction"));
+    sentry_value_set_by_key(tx, "timestamp",
+        sentry__value_new_string_owned(
+            sentry__msec_time_to_iso8601(sentry__msec_time())));
+    sentry_value_set_by_key(tx, "level", sentry_value_new_string("info"));
+
+    // TODO(tracing): add tracestate
+    // set up trace context so it mirrors the final json value
+    sentry_value_set_by_key(tx, "status", sentry_value_new_string("ok"));
+
+    sentry_value_t trace_context = sentry__span_get_trace_context(tx);
+    sentry_value_t contexts = sentry_value_new_object();
+    sentry_value_set_by_key(contexts, "trace", trace_context);
+    sentry_value_set_by_key(tx, "contexts", contexts);
+
+    // clean up trace context fields
+    sentry_value_remove_by_key(tx, "trace_id");
+    sentry_value_remove_by_key(tx, "span_id");
+    sentry_value_remove_by_key(tx, "parent_span_id");
+    sentry_value_remove_by_key(tx, "op");
+    sentry_value_remove_by_key(tx, "description");
+    sentry_value_remove_by_key(tx, "status");
+
+    // This decrefs for us, generates an event ID, merges scope
```

Comment:
```suggestion
    // This takes ownership of the transaction, generates an event ID, merges scope
```
(pretty minor, i know but i think it's clearer)